### ISSUE TITLE
[prefer-readonly-parameter-types] Add option to ignore the return type of getters

### ISSUE THREAD

**Issue by krisztianb** [typescript-eslint/typescript-eslint#2574]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
   "@typescript-eslint/prefer-readonly-parameter-types": ["error"]
  }
}
```

```TS
class Test {
    private t = { test: 1234 };

    get tt(): { test : number } {
        return this.t;
    }
}

function useTest(parameter: Readonly<Test>): void {
}
```

**tsconfig.json**

```JS
{
    "compilerOptions": {
        "lib": ["ES2015"],
        "module": "commonjs",
        "outDir": "dist",
        "strict": true,
        "target": "es2015"
    },
    "include": ["src"]
}
```

**Expected Result**

Using `Readonly<>` should be enough to mark the parameter of `useTest` as read only.

The [docs](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/prefer-readonly-parameter-types.md) say that:

> A type is considered readonly if: ... it is an object type whose properties are all marked as readonly, and whose values are all considered readonly

So it seems that getters are also checked for their return types. If I also add `Readonly<>` to the return type of the getter then the error is gone.

I can understand this behavior. But 😄 when one is using a third party library where the return types of getters are not marked as read only, then it is not possible to pass objects of those types as parameters, is it?

Since it is "not possible" to alter third party code, it would make sense to add an option to disable the checking of getter return types.

**Actual Result**

For the line with the definition of `useTest` ESLint reports:

> Parameter should be a read only type. eslint(@typescript-eslint/prefer-readonly-parameter-types)

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.1` |
| `@typescript-eslint/parser`        | `4.1.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.9.0` |
| `node`                             | `12.13.0` |

**Comment by bradzacher** [comment]
Getters are just fancy properties - they must be readonly as well.
For example in your case, if this check ignored for getters, you would easily be able to mutate the instance.

This pattern is very common for "internally read/write, externally read only" properties.

```ts
class Test {
    private t = { test: 1234 };

    get tt(): { test : number } {
        return this.t;
    }
}

function useTest(parameter: Readonly<Test>): void {
  parameter.tt.test = 1;
}

const x = new Test();
console.log(x.tt); // ==> { test: 1234 }
useTest(x);
console.log(x.tt); // ==> { test: 1 }
```

The point of the rule is to enforce types such that you _cannot_ mutate anything on the type, and such the suggested option would create a large hole in the rule's checks.

**Comment by krisztianb** [comment]
I completely agree. It would make something look read only while it isn't and thus give you some kind of false security or even worse: one would possibly overlook errors, because you figure that the error can't be in this part of the code, because the parameters are read only.

The problem I have is that I can't use the rule with third party libraries or I would have to disable it everywhere I use third party library types in the code.

Maybe a better solution would be an option like `onlyWarnOnUnsafeGetters: boolean`. The rule would then only display a warning in these cases and not an error. I don't know if it is possible to implement something like this.

**Comment by bradzacher** [comment]
Unfortunately eslint does not have the capability for a rule to change its level on the fly. Rules can only report. User config defines the severity.

**Comment by krisztianb** [comment]
Thanks. I think I learned a lot while looking into this option and the response you gave me.

I guess I just can't use this option in "error mode" then. Using it in "warning mode" might be an option. I'll think about that and closing this.
